### PR TITLE
[Sdk 514] Consent removed on changind device id without merge

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

File: countlyCommon/countlyCommon/CountlyBase.cs [hunk 1]
```diff
@@ -1017,6 +1026,12 @@ public async Task ChangeDeviceId(String newDeviceId, bool serverSideMerge = fals
         internal bool consentRequired = false;
         internal Dictionary<ConsentFeatures, bool> givenConsent = new Dictionary<ConsentFeatures, bool>();
 
+        internal enum ConsentChangedAction
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/CountlyBase.cs:1029]: Move this declaration (together with the decleration of the "ConsentFeatures" enum, "consentRequired ", "givenConsent ") to the top

File: countlyCommon/countlyCommon/CountlyBase.cs [hunk 2]
```diff
@@ -1038,7 +1053,12 @@ internal bool IsConsentGiven(ConsentFeatures feature)
 
         public async Task SetConsent(Dictionary<ConsentFeatures, bool> consentChanges)
         {
-            UtilityHelper.CountlyLogging("[CountlyBase] Calling 'SetConsent'");
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/CountlyBase.cs:1041]: It's an error to remove this log call as that removes an important piece of information in the calling chain logs

File: countlyCommon/countlyCommon/CountlyBase.cs [hunk 3]
```diff
@@ -996,14 +996,23 @@ public async Task ChangeDeviceId(String newDeviceId, bool serverSideMerge = fals
                 //if no server side merge is needed, we just end the previous session and start a new session with the new id
                 await SessionEnd();
                 await DeviceData.SetPreferredDeviceIdMethod(DeviceIdMethodInternal.developerSupplied, newDeviceId);
+                if (consentRequired) {
+                    Dictionary<ConsentFeatures, bool> removedConsent = new Dictionary<ConsentFeatures, bool>();
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/CountlyBase.cs:1000]: Rework this into an internal call to remove all consent.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Add 2 meaninigful tests to validate this functionality

**Comment by ArtursKadikis** [comment]
@ZahidZafar this has a merge conflict

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Rename "RemoveAllConsent" to "RemoveAllConsentInternal".

Add 2 meaningful tests to this change.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Major breaking changes should be at the top in the changelog.

None of the added tests are validating the changes in this PR.

**Comment by ZahidZafar** [comment]
> Major breaking changes should be at the top in the changelog.
> 
> None of the added tests are validating the changes in this PR.
**
image
**
I have posted an image in which consents are being validated on changing device id without merge.
Please let me know what kind of unit test you have in mind.

File: countlyCommon/TestingRelated/DeviceIdTests.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            SessionEvent model = Countly.Instance.Sessions[0];
+            NameValueCollection collection = HttpUtility.ParseQueryString(model.Content);
+
+            Assert.Equal("1", collection.Get("end_session"));
+            Assert.Equal(oldDeviceId, collection.Get("device_id"));
+
+            ConsentFeatures[] consents = System.Enum.GetValues(typeof(ConsentFeatures)).Cast<ConsentFeatures>().ToArray();
+            foreach (ConsentFeatures c in consents) {
+                Assert.False(Countly.Instance.IsConsentGiven(c));
+            }
```
Comment (ArtursKadikis) [inline countlyCommon/TestingRelated/DeviceIdTests.cs:107]: give consent afterwards and the verify

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**